File: teachprogramming/static/projects/tron.vb
Public Class Form1
    Dim screen As Bitmap
    Dim background_color = Color.FromArgb(0, 0, 0)

    Dim player1_x_pos As Integer  ' VER: line
    Dim player1_y_pos As Integer  ' VER: line
    Dim player1_x_move As Integer ' VER: line
    Dim player1_y_move As Integer ' VER: line
    Dim player1_color = Color.FromArgb(255, 255, 0) ' VER: line
    ' VER: line
    Dim player2_x_pos As Integer  ' VER: player2
    Dim player2_y_pos As Integer  ' VER: player2
    Dim player2_x_move As Integer ' VER: player2
    Dim player2_y_move As Integer ' VER: player2
    Dim player2_color = Color.FromArgb(255, 0, 0) ' VER: player2
    ' VER: player2
    Dim player1_score As Integer ' VER: score
    Dim player2_score As Integer ' VER: score
    ' VER: score
    Public Sub New()
        InitializeComponent()
        Timer1.Interval = Int(1000 / 60)
        Me.MaximumSize = New System.Drawing.Size(640, 480)
        Me.MinimumSize = Me.MaximumSize
        Me.PictureBox1.Size = Me.MaximumSize
        screen = New Bitmap(PictureBox1.Width, PictureBox1.Height)
        PictureBox1.Image = screen
        reset()
    End Sub

    Sub reset()
        player1_x_pos = 100                 ' VER: line
        player1_y_pos = 100                 ' VER: line
        player1_x_move = 1                  ' VER: line
        player1_y_move = 0                  ' VER: line
        player2_x_pos = screen.Width - 100  ' VER: player2
        player2_y_pos = screen.Height - 100 ' VER: player2
        player2_x_move = -1                 ' VER: player2
        player2_y_move = 0                  ' VER: player2
        ' VER: line
        Graphics.FromImage(screen).FillRectangle(New SolidBrush(background_color), 0, 0, screen.Width, screen.Height)
        Timer1.Start()
    End Sub

    Private Sub Timer1_Tick(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Timer1.Tick
        Dim g = Graphics.FromImage(screen)
        Dim pixel As Color ' VER: colide

        player1_x_pos = player1_x_pos + player1_x_move ' VER: line
        player1_y_pos = player1_y_pos + player1_y_move ' VER: line
        player2_x_pos = player2_x_pos + player2_x_move ' VER: player2
        player2_y_pos = player2_y_pos + player2_y_move ' VER: player2
        ' VER: line
        Try                                                           ' VER: colide
            pixel = screen.GetPixel(player1_x_pos, player1_y_pos)     ' VER: colide
        Catch ex As Exception                                         ' VER: colide
            pixel = Color.White                                       ' VER: colide
        End Try                                                       ' VER: colide
        If Not background_color.Equals(pixel) Then                    ' VER: colide
            Timer1.Stop()                                             ' VER: score
            player2_score = player2_score + 1                         ' VER: score
            MsgBox("Player 1 Died " + player1_score.ToString)         ' VER: score
            reset()                                                   ' VER: colide
        End If                                                        ' VER: colide
        ' VER: colide
        Try                                                           ' VER: player2
            pixel = screen.GetPixel(player2_x_pos, player2_y_pos)     ' VER: player2
        Catch ex As Exception                                         ' VER: player2
            pixel = Color.White                                       ' VER: player2
        End Try                                                       ' VER: player2
        If Not background_color.Equals(pixel) Then                    ' VER: player2
            Timer1.Stop()                                             ' VER: score
            player1_score = player1_score + 1                         ' VER: score
            MsgBox("Player 2 Died " + player2_score.ToString)         ' VER: score
            reset()                                                   ' VER: player2
        End If                                                        ' VER: player2
        ' VER: player2
        screen.SetPixel(player1_x_pos, player1_y_pos, player1_color) ' VER: line
        screen.SetPixel(player2_x_pos, player2_y_pos, player2_color) ' VER: player2
        ' VER: line
        If player1_score = 5 Or player2_score = 5 Then ' VER: score
            Me.Close()                                 ' VER: score
        End If                                         ' VER: score
        PictureBox1.Refresh()
    End Sub

    Private Sub Form1_KeyDown(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles Me.KeyDown ' VER: input
        If e.KeyCode = Keys.Up Then    ' VER: input
            player1_x_move = 0         ' VER: input
            player1_y_move = -1        ' VER: input
        End If                         ' VER: input
        If e.KeyCode = Keys.Down Then  ' VER: input
            player1_x_move = 0         ' VER: input HIDE
            player1_y_move = 1         ' VER: input HIDE
        End If                         ' VER: input HIDE
        If e.KeyCode = Keys.Left Then  ' VER: input HIDE
            player1_x_move = -1        ' VER: input HIDE
            player1_y_move = 0         ' VER: input HIDE
        End If                         ' VER: input HIDE
        If e.KeyCode = Keys.Right Then ' VER: input HIDE
            player1_x_move = 1         ' VER: input HIDE
            player1_y_move = 0         ' VER: input HIDE
        End If                         ' VER: input HIDE
        ' VER: input
        If e.KeyCode = Keys.W Then     ' VER: player2
            player2_x_move = 0         ' VER: player2
            player2_y_move = -1        ' VER: player2
        End If ' VER: player2
        If e.KeyCode = Keys.S Then     ' VER: player2
            player2_x_move = 0         ' VER: player2 HIDE
            player2_y_move = 1         ' VER: player2 HIDE
        End If                         ' VER: player2 HIDE
        If e.KeyCode = Keys.A Then     ' VER: player2 HIDE
            player2_x_move = -1        ' VER: player2 HIDE
            player2_y_move = 0         ' VER: player2 HIDE
        End If                         ' VER: player2 HIDE
        If e.KeyCode = Keys.D Then     ' VER: player2 HIDE
            player2_x_move = 1         ' VER: player2 HIDE
            player2_y_move = 0         ' VER: player2 HIDE
        End If                         ' VER: player2 HIDE
        ' VER: player2
    End Sub                            ' VER: input
    ' VER: input
End Class

File: teachprogramming/static/projects/tron.py
import pygame

pygame.init()
screen = pygame.display.set_mode((640, 480))
clock  = pygame.time.Clock()

class callByRef:
    def __init__(self, **args):
        for (key, value) in args.items():
            setattr(self, key, value)
variables = callByRef(
    running           = True,
    background_colour = (  0,   0,   0, 255),
    color_exit        = (255,   0,   0, 255), # VER: maze
    level_number      = 1,                    # Ver: maze
    player1_color     = (255, 255,   0     ), # Ver: line
    player1_x_pos     = None,                 # Ver: line
    player1_y_pos     = None,                 # Ver: line
    player1_x_move    = None,                 # Ver: line
    player1_y_move    = None,                 # Ver: line
    player2_color     = (255,   0,   0     ), # Ver: player2
    player2_x_pos     = None,                 # Ver: player2
    player2_y_pos     = None,                 # Ver: player2
    player2_x_move    = None,                 # Ver: player2
    player2_y_move    = None,                 # Ver: player2
    player1_score     = 0,                    # Ver: score
    player2_score     = 0,                    # Ver: score
)

def reset():
    pygame.draw.rect(screen, variables.background_colour, pygame.Rect(0, 0, screen.get_width(), screen.get_height()))
    #level_image = pygame.image.load("MazeLevel%d.gif" % level_number) # VER: maze
    #screen.blit(level_image, level_image.get_rect())                  # VER: maze
    variables.player1_x_pos    = 100 # Ver: line
    variables.player1_y_pos    = 100 # Ver: line
    variables.player1_x_move   =   1 # Ver: line
    variables.player1_y_move   =   0 # Ver: line
    variables.player2_x_pos    = screen.get_width()  - 100 # Ver: player2
    variables.player2_y_pos    = screen.get_height() - 100 # Ver: player2
    variables.player2_x_move   =  -1                       # Ver: player2
    variables.player2_y_move   =   0                       # Ver: player2

reset()
while variables.running:
    clock.tick(60)
    
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            variables.running = False
    
    keys = pygame.key.get_pressed()   # Ver: input
    if keys[pygame.K_ESCAPE]:         # Ver: input
        variables.running = False     # Ver: input
                                      # Ver: input
    if keys[pygame.K_UP    ]:         # Ver: input
        variables.player1_x_move =  0 # Ver: input
        variables.player1_y_move = -1 # Ver: input
    if keys[pygame.K_RIGHT ]:         # Ver: input
        variables.player1_x_move =  1 # Ver: input HIDE
        variables.player1_y_move =  0 # Ver: input HIDE
    if keys[pygame.K_LEFT  ]:         # Ver: input HIDE
        variables.player1_x_move = -1 # Ver: input HIDE
        variables.player1_y_move =  0 # Ver: input HIDE
    if keys[pygame.K_DOWN  ]:         # Ver: input HIDE
        variables.player1_x_move =  0 # Ver: input HIDE
        variables.player1_y_move =  1 # Ver: input HIDE
                                      # Ver: player2
    if keys[pygame.K_w     ]:         # Ver: player2
        variables.player2_x_move =  0 # Ver: player2
        variables.player2_y_move = -1 # Ver: player2
    if keys[pygame.K_d     ]:         # Ver: player2
        variables.player2_x_move =  1 # Ver: player2 HIDE
        variables.player2_y_move =  0 # Ver: player2 HIDE
    if keys[pygame.K_a     ]:         # Ver: player2 HIDE
        variables.player2_x_move = -1 # Ver: player2 HIDE
        variables.player2_y_move =  0 # Ver: player2 HIDE
    if keys[pygame.K_s     ]:         # Ver: player2 HIDE
        variables.player2_x_move =  0 # Ver: player2 HIDE
        variables.player2_y_move =  1 # Ver: player2 HIDE
                                      # Ver: player2
    variables.player1_x_pos = variables.player1_x_pos + variables.player1_x_move # Ver: line
    variables.player1_y_pos = variables.player1_y_pos + variables.player1_y_move # Ver: line
                                                                                 # Ver: line
    variables.player2_x_pos = variables.player2_x_pos + variables.player2_x_move # Ver: player2
    variables.player2_y_pos = variables.player2_y_pos + variables.player2_y_move # Ver: player2 HIDE
                                                                                 # Ver: player2
    if variables.player1_x_pos<=0                  : variables.player1_x_pos = screen.get_width() -1 # VER: wrap
    if variables.player1_x_pos>=screen.get_width() : variables.player1_x_pos =                     1 # VER: wrap
    if variables.player1_y_pos<=0                  : variables.player1_y_pos = screen.get_height()-1 # VER: wrap HIDE
    if variables.player1_y_pos>=screen.get_height(): variables.player1_y_pos =                     1 # VER: wrap HIDE
    if variables.player2_x_pos<=0                  : variables.player2_x_pos = screen.get_width() -1 # VER: player2,wrap
    if variables.player2_x_pos>=screen.get_width() : variables.player2_x_pos =                     1 # VER: player2,wrap
    if variables.player2_y_pos<=0                  : variables.player2_y_pos = screen.get_height()-1 # VER: player2,wrap HIDE
    if variables.player2_y_pos>=screen.get_height(): variables.player2_y_pos =                     1 # VER: player2,wrap HIDE
                                                                                                     # VER: wrap
    try   : player1_at_pixel = screen.get_at((variables.player1_x_pos, variables.player1_y_pos)) # Ver: colide
    except: player1_at_pixel = None                                                              # Ver: colide
    if player1_at_pixel == variables.color_exit:                                                 # VER: maze
        variables.level_number += 1                                                              # VER: maze
    if player1_at_pixel != variables.background_colour:                                          # Ver: colide
        variables.player2_score = variables.player2_score + 1                                    # Ver: score
        reset()                                                                                  # Ver: colide
                                                                                                 # Ver: colide
    try   : player2_at_pixel = screen.get_at((variables.player2_x_pos, variables.player2_y_pos)) # Ver: player2
    except: player2_at_pixel = None                                                              # Ver: player2 HIDE
    if player2_at_pixel != variables.background_colour:                                          # Ver: player2
        variables.player1_score = variables.player1_score + 1                                    # Ver: score HIDE
        reset()                                                                                  # Ver: player2 HIDE
                                                                                                 # Ver: player2
    screen.set_at((variables.player1_x_pos, variables.player1_y_pos), variables.player1_color) # Ver: line
    screen.set_at((variables.player2_x_pos, variables.player2_y_pos), variables.player2_color) # Ver: player2 HIDE
                                                                                               # Ver: line
    if variables.player1_score == 5 or variables.player2_score == 5: # Ver: score
        variables.running = False                                    # Ver: score
                                                                     # Ver: score
    pygame.display.flip()

pygame.quit()
                               # Ver: score
print("Player 1 Score")        # Ver: score
print(variables.player1_score) # Ver: score
print("Player 2 Score")        # Ver: score
print(variables.player2_score) # Ver: score